sage/prove_group_implementations.sage

```diff
@@ -80,43 +77,42 @@ def formula_secp256k1_gej_add_ge_var(branch, a, b):
   s2 = s2 * a.Z
   h = -u1
   h = h + u2
-  i = -s1
-  i = i + s2
+  i = -s2
+  i = i + s1
   if (branch == 2):
     r = formula_secp256k1_gej_double_var(a)
     return (constraints(zero={b.Z - 1 : 'b.z=1'}), constraints(zero={a.Infinity : 'a_finite', b.Infinity : 'b_finite', h : 'h=0', i : 'i=0'}), r)
   if (branch == 3):
     return (constraints(zero={b.Z - 1 : 'b.z=1'}), constraints(zero={a.Infinity : 'a_finite', b.Infinity : 'b_finite', h : 'h=0'}, nonzero={i : 'i!=0'}), point_at_infinity())
-  i2 = i^2
-  h2 = h^2
-  h3 = h * h2
   rz = a.Z * h
+  h2 = h^2
+  h2 = -h2
+  h3 = h2 * h
   t = u1 * h2
-  rx = t
-  rx = rx * 2
+  rx = i^2
   rx = rx + h3
-  rx = -rx
-  rx = rx + i2
-  ry = -rx
-  ry = ry + t
-  ry = ry * i
+  rx = rx + t
+  rx = rx + t
+  t = t + rx
+  ry = t * i
   h3 = h3 * s1
-  h3 = -h3
   ry = ry + h3
   return (constraints(zero={b.Z - 1 : 'b.z=1'}), constraints(zero={a.Infinity : 'a_finite', b.Infinity : 'b_finite'}, nonzero={h : 'h!=0'}), jacobianpoint(rx, ry, rz))
 
 def formula_secp256k1_gej_add_zinv_var(branch, a, b):
   """libsecp256k1's secp256k1_gej_add_zinv_var"""
   bzinv = b.Z^(-1)
-  if branch == 0:
-    return (constraints(), constraints(nonzero={b.Infinity : 'b_infinite'}), a)
-  if branch == 1:
+  if branch == 0 or branch == 1:
     bzinv2 = bzinv^2
     bzinv3 = bzinv2 * bzinv
     rx = b.X * bzinv2
     ry = b.Y * bzinv3
     rz = 1
+    if branch == 0:
+      return (constraints(), constraints(nonzero={a.Infinity : 'a_infinite', b.Infinity : 'b_infinite'}), point_at_infinity())
```

Comment:
Why is this branch inside this block `branch == 0 or branch == 1` and not just before? It doesn't use any of the local variables. Also, why does it exist at all, given that there's no corresponding branch in the C code?

Reply:
This branch corresponds to https://github.com/bitcoin-core/secp256k1/blob/2d0e1154c1939e8e7ffb05f64faf3f626fcb7000/src/group_impl.h#L455 in the C code, which is equivalent to
```
if (b->infinity) {
    r ->infinity = 1
} else {
    b ->infinity = 1
}
```

So the mapping then is  
 * `(C) a->infinity && b->infinity` -> `(sage) branch == 0`
 * `(C) a->infinity && !b->infinity` -> `(sage) branch == 1`

When I wrote this I assumed that in sage, we need to write it as an explicit branch because we have no infinity field but instead `point_at_infinity()` and `jacobianpoint(x, y, z)` constructors. But now I see that it's not entirely true. The constructor is actually [`jacobianpoint(x, y, z, infinity)`](https://github.com/bitcoin-core/secp256k1/blob/master/sage/weierstrass_prover.sage#L18), so there's a much simpler way that's closer to the sage code, see
https://github.com/real-or-random/secp256k1/commit/0946486cec44c99befe89fbb6ae5be8f48e33dcf .

@peterdettman Can you pick that commit and squash again?

Reply:
> `b ->infinity = 1`

You mean `r ->infinity = 0`? Anyway, I see what you meant to represent with this branch. Your fixup looks good.

Reply:
yes... edited my comment to fix.